### PR TITLE
Add movie suggestions to movie_details using the KNN recommendation model 

### LINKED ISSUE #31: Add suggested movies to movie_details page
While on the movie details_page implement a feature to show 5 related movies. Use the KNN code for the implementation. If the movie is not present in the db show a message saying that movie is not currently available.

### PR DESCRIPTION
**Pull request by Aryan10**
## Description

Integrated the KNN-based recommendation system that suggests similar movies to a particular move based on genre with the web page to show recommendations. This is a revised version of PR #33

The new approach differs a bit as suggested by #31 Instead of searching for movie name in the local dataset, it uses the genres generated by API, maps them to genres in the dataset to use the KNN model for predicitons, this makes suggestions immediately avaliable for all movies. Further, posters of those movies are searched from the OMDB API and displayed on the movie_details page.

## Related Issue

Fixes #31 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Motivation and Context

This feature allows users to see movie recommendations based on the current movie they’re viewing, enhancing user engagement.

## How Has This Been Tested?

The feature was manually tested in the development environment to verify:

- The similarity recommendations display correctly for each movie
- No errors occur when movies with no recommendations are encountered
- Template renders without layout or display issues
- Posters are fetched and display correctly from the OMDB api

## Screenshots:
<details>
<summary>
Recommendations for Interstellar

Genre: Adventure, Drama, Sci-Fi
</summary>

image

image

</details>

<details>
<summary>
Recommendations for Shrek

Genre: Animation, Adventure, Comedy
</summary>

image

image

</details>

## Checklist:

- [x] I have registered myself at [Contrihub website](https://sac.mnnit.ac.in/contrihub).
- [x] My code follows the code style of this project.
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] Any dependent changes have been merged and published in downstream modules
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by Aryan10** [comment]
@eleensmathew Please verify if this contains everything.

**Comment by eleensmathew** [comment]
Lgtm but i still need to check on my device. I'll review the pr and check it out

**Comment by Aryan10** [comment]
Also ma'am you might need to reject the old PR on ContriHUB site so I submit another one (can not request assignment atm).

**Comment by eleensmathew** [comment]
No need to request, i'll just review this pr and after merging I'll just accept the pr submitted on the website

**Comment by Aryan10** [comment]
Ok got it.

File: notebook/recommendation.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[81]:
```
Comment (eleensmathew) [inline notebook/recommendation.py]: remove these comments

File: notebook/recommendation.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import os
+
+BASE_DIR = os.path.dirname(os.path.abspath(__file__))
+
+paths = {
+    "movies.csv": os.path.join(BASE_DIR, "data/movies.csv"),
+    "ratings.csv": os.path.join(BASE_DIR, "data/ratings.csv")
+}
+
+# In[82]:
```
Comment (eleensmathew) [inline notebook/recommendation.py]: remove these comments

File: notebook/recommendation.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+# ratings = pd.read_csv(paths["ratings.csv"])
+movies = pd.read_csv(paths["movies.csv"])
+
+print("Datasets loaded")
+
+
+# In[84]:
+
+
+(movies.shape)
```
Comment (eleensmathew) [inline notebook/recommendation.py]: these lines are not required here

File: notebook/recommendation.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+data = movies
+
+# one-hot encoding
+mlb = MultiLabelBinarizer()
+genre_encoded = mlb.fit_transform(data['genres'].str.split('|'))
+genre_df = pd.DataFrame(genre_encoded, columns=mlb.classes_)
+
+data = pd.concat([data, genre_df], axis=1)
+data.head()
```
Comment (eleensmathew) [inline notebook/recommendation.py]: not required here

File: notebook/recommendation.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+# In[139]:
+
+
+def recommend_similar_movies_model1(movie_title, *, n=5, model=knn_model_genres, data=movies, features=features):
+    movie_idx = data[data['title'] == movie_title].index[0]
+    distances, indices = model.kneighbors(features.iloc[movie_idx, :].values.reshape(1, -1), n_neighbors=n)
+    recommended_movies = data.iloc[indices.flatten()]
+    return recommended_movies
+
+recommend_similar_movies_model1("Toy Story (1995)")
```
Comment (eleensmathew) [inline notebook/recommendation.py]: remove this line

**Review by eleensmathew — CHANGES_REQUESTED** [review]

**Comment by Aryan10** [comment]
I have removed those comments and also pushed some code in styles.css in line with the original design

image

**Comment by eleensmathew** [comment]
Can you change it so that if a picture isnt found a default pic is shown?

**Comment by Aryan10** [comment]
@eleensmathew Can you give me an example where this happens so I can test it?

**Comment by Aryan10** [comment]
Ok wait I was able to find an example, I will add a default image for when no image is found.

**Comment by Aryan10** [comment]
@eleensmathew I have added the placeholder image, please take a look

**Comment by Aryan10** [comment]
image

**Comment by Aryan10** [comment]
@eleensmathew Have you looked into this? Or do I need to make any more changes.

**Comment by eleensmathew** [comment]
No need for more changes

**Comment by eleensmathew** [comment]
The placeholder image works fine but the styling in the web page appears like this rn to me 
image

I think you have added styling here right?

**Comment by eleensmathew** [comment]
Also can you resolve the conflicts

**Comment by Aryan10** [comment]
Ok I will resolve the conflicts, I think its mostly just import statements wait

**Comment by Aryan10** [comment]
@eleensmathew I have resolved the merge conflicts and tested the site

**Comment by Aryan10** [comment]
@eleensmathew for the styling, it looks correct to me, have you tried reloading with Ctrl+F5?

**Comment by eleensmathew** [comment]
I'll check it again

**Comment by Aryan10** [comment]
@eleensmathew Ok wait I will also re check, I noticed the structure of css was changed a bit

**Comment by eleensmathew** [comment]
Hey it looks good now. I will merge it

**Pull request merged**